pkg/cluster/stores.go

```diff
@@ -21,155 +19,100 @@ import (
 	}
 }
 
 // Update the store set to the new set of addresses and labels for that addresses.
 // New background processes initiated respect the lifecycle of the given context.
 func (s *StoreSet) Update(ctx context.Context) {
-	// XXX(fabxc): The store as is barely ties into the cluster. This is the only place where
-	// we depend on it. However, in the future this may change when we fetch additional information
-	// about storePeers or make the set self-updating through events rather than explicit calls to Update.
-	storePeers := map[string]PeerState{}
-	for _, ps := range s.peer.PeerStates(PeerTypesStoreAPIs()...) {
-		storePeers[ps.APIAddr] = ps
-	}
+	var err error
 
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
+	stores := make([]*query.StoreInfo, 0, len(s.conns))
+	conns := make(map[string]*grpc.ClientConn, len(s.conns))
 
-	// For each new store peer we create a new gRPC connection.
-	// In every case we also updates the labels from peer state.
-	for addr, state := range storePeers {
-		if _, ok := s.stores[addr]; !ok {
-			level.Debug(s.logger).Log("msg", "grpc dialing", "store", addr)
-
-			startTime := time.Now()
-			conn, err := grpc.DialContext(ctx, addr,
-				grpc.WithInsecure(),
-				grpc.WithUnaryInterceptor(
-					grpc_middleware.ChainUnaryClient(
-						s.grpcMetrics.UnaryClientInterceptor(),
-						tracing.UnaryClientInterceptor(s.tracer),
-					),
-				),
-				grpc.WithStreamInterceptor(
-					grpc_middleware.ChainStreamClient(
-						s.grpcMetrics.StreamClientInterceptor(),
-						tracing.StreamClientInterceptor(s.tracer),
-					),
-				),
-			)
+	for _, ps := range s.peer.PeerStates(PeerTypesStoreAPIs()...) {
+		conn, ok := s.conns[ps.APIAddr]
+		if !ok {
+			conn, err = grpc.DialContext(ctx, ps.APIAddr, s.dialOpts...)
 			if err != nil {
-				s.storeNodeFailedDials.Inc()
-				level.Warn(s.logger).Log("msg", "dialing connection failed; skipping", "store", addr, "err", err)
+				level.Warn(s.logger).Log("msg", "dialing connection failed; skipping",
+					"store", ps.APIAddr, "err", err)
 				continue
 			}
-
-			s.storeNodeDialDuration.Observe(time.Since(startTime).Seconds())
-			level.Debug(s.logger).Log("msg", "successfully made grpc connection", "store", addr)
-
-			store := &storeInfo{conn: conn}
-			s.stores[addr] = store
 		}
-
-		// Always fetch up-to-date labels propagated in peer state.
-		s.stores[addr].setLabels(state.Metadata.Labels)
+		stores = append(stores, &query.StoreInfo{
+			Addr:    ps.APIAddr,
+			Client:  storepb.NewStoreClient(conn),
+			Labels:  ps.Metadata.Labels,
+			MinTime: ps.Metadata.MinTime,
+			MaxTime: ps.Metadata.MaxTime,
+		})
+		conns[ps.APIAddr] = conn
 	}
 
-	// Delete stores that no longer exist.
-	for addr, store := range s.stores {
-		if _, ok := storePeers[addr]; !ok {
-			store.conn.Close()
-			level.Info(s.logger).Log("msg", "closing connection for store")
-			delete(s.stores, addr)
-		}
-	}
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.stores = stores
+	s.conns = conns
```

Comment:
Are we leaking connections here? You are creating new slice with connections without these not presented by current `s.peer.PeerStates`. So I think we are leaking grpc connections (and go routines) here.

Reply:
They'd probably get killed of eventually – but you are right, we should do this properly.